### PR TITLE
Add `MaskedInput` component

### PR DESCRIPTION
**Pull request by alex-ju**
### :pushpin: Summary

A form field (`<input>` or `<textarea>`) that obfuscates characters to protect sensitive input/information by replacing them with a shape.

### :hammer_and_wrench: Detailed description

The PR is currently in a draft state as we're exploring the most suitable implementation of this component.

The proposed implementation is built using the `Form::TextInput::Base` and `Form::Textarea::Base`, inheriting most of their API, with the following additions:
 - `@isMasked` (default `true`) allow consumers to turn the masking off (if required) as well as controlling it from outside the component
 - `@isMultiline` (default `false`) used to render a `<textarea>` element instead of an `<input>` (an alternative to this argument would be to have a separate `Form::MaskedTextarea` component)

Feedback welcome! 🙏 

[👉 Showcase preview](https://hds-showcase-git-alex-ju-add-sensitive-input-hashicorp.vercel.app/components/form/masked-input)

### :link: External links

Jira ticket: [HDS-2115](https://hashicorp.atlassian.net/browse/HDS-2115)
Figma file: https://www.figma.com/file/zqGYaQb6JjTM8FuZoXYd9a/Sensitive-input?type=design&node-id=38284-71082&mode=dev

***

### 👀 Reviewer's checklist:

- [x] +1 Percy if applicable
- [x] Confirm that PR has a changelog update via [Changesets](https://github.com/changesets/changesets) if needed

:speech_balloon: Please consider using [conventional comments](https://conventionalcomments.org/) when reviewing this PR.

[HDS-2115]: https://hashicorp.atlassian.net/browse/HDS-2115?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by heatherlarsen — COMMENTED** [review]
This looks good, @alex-ju. I'll let @cveigt do a more thorough review of the designs when he's back tomorrow. 

To your question about having a separate Textarea component vs `isMultiline`, I don't have strong opinions on this. I think both are fairly straightforward.

File: packages/components/addon/components/hds/form/sensitive-input/field.hbs
Comment (didoo) [inline packages/components/addon/components/hds/form/sensitive-input/field.hbs:21]: the `@height` should be forwarded too

File: packages/components/app/styles/components/form/sensitive-input.scss
Comment (didoo) [inline packages/components/app/styles/components/form/sensitive-input.scss:36]: why the difference between these two? 🤔

File: packages/components/tests/dummy/app/templates/components/form/sensitive-input.hbs
Comment (didoo) [inline packages/components/tests/dummy/app/templates/components/form/sensitive-input.hbs:285]: One thing I noticed, is that in the `Textarea` we don't test the `@height` argument. Maybe we could do it this context (here and also in the `Textarea`)?

File: packages/components/tests/integration/components/hds/form/sensitive-input/field-test.js
Comment (didoo) [inline packages/components/tests/integration/components/hds/form/sensitive-input/field-test.js:43]: should we test also the height for the multiline version?

**Review by didoo — COMMENTED** [review]
@alex-ju added a few comments. overall the API makes sense to me.

regarding the component name, no strong opinions: `SensitiveInput` works, as well as `ObfuscatedInput` too

File: packages/components/addon/components/hds/form/sensitive-input/base.hbs
Comment (alex-ju) [reply packages/components/addon/components/hds/form/sensitive-input/base.hbs]: Indeed, removed now in [be7de9c](https://github.com/hashicorp/design-system/pull/1468/commits/be7de9c08c7f883350a0163fd50f9236394f290f)

File: packages/components/addon/components/hds/form/sensitive-input/field.hbs
Comment (alex-ju) [reply packages/components/addon/components/hds/form/sensitive-input/field.hbs:21]: Great spot! Fixed in [6361cd8](https://github.com/hashicorp/design-system/pull/1468/commits/6361cd83e7bf6fd349a1df8a1fbce220319a109a)

File: packages/components/app/styles/components/form/sensitive-input.scss
Comment (alex-ju) [reply packages/components/app/styles/components/form/sensitive-input.scss]: Subtracting `1px` for the TextInput/Textarea border (as the icon is positioned relative to the wrapping div, not the input). It should've been applied to the right as well. Updated in [7d5ebb9](https://github.com/hashicorp/design-system/pull/1468/commits/7d5ebb916aabcc7b69dd6d77a71061e472677b7e) and used tokens to make the computation more clear.

File: packages/components/tests/dummy/app/templates/components/form/sensitive-input.hbs
Comment (alex-ju) [reply packages/components/tests/dummy/app/templates/components/form/sensitive-input.hbs:285]: Added in [6361cd8](https://github.com/hashicorp/design-system/pull/1468/commits/6361cd83e7bf6fd349a1df8a1fbce220319a109a)

File: packages/components/tests/integration/components/hds/form/sensitive-input/field-test.js
Comment (alex-ju) [reply packages/components/tests/integration/components/hds/form/sensitive-input/field-test.js:43]: Added in [6361cd8](https://github.com/hashicorp/design-system/pull/1468/commits/6361cd83e7bf6fd349a1df8a1fbce220319a109a)

File: packages/components/addon/components/hds/form/sensitive-input/base.hbs
Comment (didoo) [inline packages/components/addon/components/hds/form/sensitive-input/base.hbs:11]: could it be that we have to forward the `@width` (for both of them)?

**Review by didoo — COMMENTED** [review]
Re-reviewed.

File: packages/components/addon/components/hds/form/sensitive-input/base.hbs
Comment (alex-ju) [reply packages/components/addon/components/hds/form/sensitive-input/base.hbs:11]: The width would have to be set at the wrapper level (see L5) as the relative element, otherwise the icon may fall outside the input.

Comment (didoo) [reply packages/components/addon/components/hds/form/sensitive-input/base.hbs:11]: Ahhh yes, of course! (I thought this was the "field" but it's not, it's the "base" of the sensitive input. Makes sense.

**Comment by cveigt** [comment]
This is looking great, Alex.

One small thing. For the read-only variants, the textarea (multiline) seems to be taking the foreground color token for disabled and not for read-only.

Screen Shot 2023-07-04 at 11 48 05 PM

Regarding having a separate Textarea component vs isMultiline. Both approaches are achievable in Figma, and I don't see significant benefits or downsides with any of them.

However, from a consumer's perspective, when designers (and even developers) need a specific form element component, they probably consider whether they need a text input or textarea element and choose/grab a specific instance (TextInput or Textarea) intentionally, and not based on a property.

If building to separate components means we're expanding the scope, I'd say let's build a single component. If it involves creating an extra documentation page and not much more time and effort. It might be worth asking ambassadors what they'd prefer.

**Comment by alex-ju** [comment]
> One small thing. For the read-only variants, the textarea (multiline) seems to be taking the foreground color token for disabled and not for read-only.
> 
> Screen Shot 2023-07-04 at 11 48 05 PM
> Regarding having a separate Textarea component vs isMultiline. Both approaches are achievable in Figma, and I don't see significant benefits or downsides with any of them.
> 
Thanks, Christian! I had the same realization when reviewing the component. As this is an issue with the `Hds::Textarea` I addressed it separately in #1470, once that is merged it will be reflected in this component as well.

**Comment by MelSumner** [comment]
We should make sure to mention in the documentation for this that it's visual obfuscation only, and that consumers should be aware that the value _could_ still be obtained from the terminal 

e.g.
CleanShot 2023-07-05 at 09 51 27

**Comment by didoo** [comment]
> We should make sure to mention in the documentation for this that it's visual obfuscation only, and that consumers should be aware that the value _could_ still be obtained from the terminal

@MelSumner added in the documentation PR: https://github.com/hashicorp/design-system/pull/1476/commits/15168cec83a09a867e0e8e7a00df61f8e85c07eb

**Review by didoo — APPROVED** [review]
👍

**Comment by didoo** [comment]
@cveigt can you please re-review and if OK approve this PR? thanks

**Review by KristinLBradley — COMMENTED** [review]
Looking great! I don't have any specific feedback at this point.

**Review by cveigt — APPROVED** [review]
🚀🚀🚀

**Pull request merged**
